pilot/pkg/networking/v1alpha3/gateway.go

```diff
@@ -16,161 +16,168 @@ package v1alpha3
 
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/route"
+
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"github.com/gogo/protobuf/types"
+
 	networking "istio.io/api/networking/v1alpha3"
 	"istio.io/istio/pilot/pkg/model"
-	deprecated "istio.io/istio/pilot/pkg/proxy/envoy/v1"
 	"istio.io/istio/pkg/log"
 )
 
-func buildGatewayHTTPListeners(env model.Environment, node model.Proxy) ([]*xdsapi.Listener, error) {
+func buildGatewayListeners(env model.Environment, node model.Proxy) ([]*xdsapi.Listener, error) {
 
 	config := env.IstioConfigStore
 
-	gateways, err := config.List(model.Gateway.Type, model.NamespaceAll)
+	// collect workload labels
+	workloadInstances, err := env.GetProxyServiceInstances(node)
 	if err != nil {
-		return nil, fmt.Errorf("listing gateways: %s", err)
+		log.Error("Failed to get gateway instances for router ", zap.String("node", node.ID), zap.Error(err))
+		return nil, err
 	}
 
+	var workloadLabels model.LabelsCollection
+	for _, w := range workloadInstances {
+		workloadLabels = append(workloadLabels, w.Labels)
+	}
+
+	gateways := config.Gateways(workloadLabels)
+
 	if len(gateways) == 0 {
 		log.Debug("no gateways for router", zap.String("node", node.ID))
 		return []*xdsapi.Listener{}, nil
 	}
 
-	gateway := &networking.Gateway{}
-	for _, spec := range gateways {
-		err := model.MergeGateways(gateway, spec.Spec.(*networking.Gateway))
-		if err != nil {
-			return nil, fmt.Errorf("merge gateways: %s", err)
-		}
+	// TODO: merging makes code simpler but we lose gateway names that are needed to determine
+	// the virtual services pinned to each gateway
+	//gateway := &networking.Gateway{}
+	//for _, spec := range gateways {
+	//	err := model.MergeGateways(gateway, spec.Spec.(*networking.Gateway))
+	//	if err != nil {
+	//		log.Errorf("Failed to merge gateway %s for router %s", spec.Name, node.ID)
+	//		return nil, fmt.Errorf("merge gateways: %s", err)
+	//	}
+	//}
+
+	// HACK for the above case
+	if len(gateways) > 1 {
+		log.Warn("Currently, Istio cannot bind multiple gateways to the same workload")
+		return []*xdsapi.Listener{}, nil
 	}
 
+	name := gateways[0].Name
+	gateway := gateways[0].Spec.(*networking.Gateway)
+
 	listeners := make([]*xdsapi.Listener, 0, len(gateway.Servers))
+	listenerPortMap := make(map[uint32]bool)
 	for _, server := range gateway.Servers {
-		// TODO: TCP
 
-		// build physical listener
-		physicalListener := buildGatewayListener(env, node, server)
-		if physicalListener == nil {
-			continue // TODO: add support for all protocols
+		// TODO: this does not handle the case where there are two servers on same port
+		if listenerPortMap[server.Port.Number] {
+			log.Warnf("Multiple servers on same port is not supported yet, port %d", server.Port.Number)
+			continue
+		}
+		switch model.Protocol(server.Port.Protocol) {
+		case model.ProtocolHTTP, model.ProtocolHTTP2, model.ProtocolGRPC, model.ProtocolHTTPS:
```

Comment:
use `server.Port.Protocol.IsHTTP()` instead?

Reply:
this is not in model.Protocol format